akka-persistence-testkit/src/main/scala/akka/persistence/testkit/InMemStorage.scala

```diff
@@ -0,0 +1,604 @@
+/*
+ * Copyright (C) 2018-2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.persistence.testkit
+
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.atomic.AtomicReference
+import java.util.{ List => JList }
+
+import akka.util.ccompat.JavaConverters._
+import akka.actor.{ ActorSystem, ExtendedActorSystem, Extension, ExtensionId, ExtensionIdProvider }
+import akka.annotation.InternalApi
+import akka.persistence.testkit.ProcessingPolicy._
+import akka.persistence._
+import akka.serialization.{ Serialization, SerializationExtension, Serializers }
+import akka.persistence.testkit.scaladsl.{ PersistenceTestKit, SnapshotTestKit }
+
+import scala.annotation.tailrec
+import scala.collection.immutable
+import scala.util.{ Failure, Success, Try }
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+sealed trait InternalReprSupport[R] {
+
+  type InternalRepr
+
+  private[testkit] def toInternal(repr: R): InternalRepr
+
+  private[testkit] def toRepr(internal: InternalRepr): R
+
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+sealed trait InMemStorage[K, R] extends InternalReprSupport[R] {
+
+  import akka.persistence.testkit.Utils.JavaFuncConversions._
+
+  private val parallelism = Runtime.getRuntime.availableProcessors()
+
+  private final val eventsMap: ConcurrentHashMap[K, Vector[InternalRepr]] =
+    new ConcurrentHashMap()
+
+  def forEachKey(f: K => Unit): Unit = eventsMap.forEachKey(parallelism, f)
+
+  def findMany(key: K, fromInclusive: Int, maxNum: Int): Option[Vector[R]] =
+    read(key).flatMap(
+      value =>
+        if (value.size > fromInclusive)
+          Some(value.drop(fromInclusive).take(maxNum))
+        else None)
+
+  def findOneByIndex(key: K, index: Int): Option[R] =
+    Option(eventsMap.get(key)).flatMap(value => if (value.size > index) Some(value(index)) else None).map(toRepr)
+
+  def add(key: K, p: R): Unit =
+    add(key, List(p))
+
+  /**
+   *
+   * Note! `elems` is call by name to preserve atomicity
```

Comment:
> it's ok to have some `@InternalApi` in the other public files also, but when it's all internal it's good to make it clear by placing it in `internal` package

@patriknw it is not 100% internal. I have put the `JournalOperation` and `SnapshotOperation` and their impls. also into this file. These messages are used in the policies to react on the type of operation performed on the storage. I don't think these messages should be in the internal package, because they can be used by external developers. 
Should I put them in a different file, or just leave `InMemStorage` in the testkit package?

Reply:
> what does atomicity mean here? why is it needed if elems is immutable?

Maybe not the best comment, I'll change. I use mapAny to obtain the elems in the implemenation:
https://github.com/akka/akka/blob/ee2d55ba04f987240d12bd4a908b4ce3759059de/akka-persistence-testkit/src/main/scala/akka/persistence/testkit/InMemStorage.scala#L226-L227

This mapAny should be invoked inside the `compute` method  of `eventsMap` to preserve the atomicity of mutations on the maps of `seqNumbers` and `eventsMap`.
I think I will rework this part, to make it clearer.